feat: wip bank guide (#252)
Basis-Theory/developers.basistheory.com: docs/guides/banks/collect-bank-accounts.mdx

import Tabs from "@theme/Tabs";
import TabItem from "@theme/TabItem";

import { Alert } from "@site/src/components/shared/Alert";
import GettingStartedSection from "@site/src/components/docs/_getting-started-section.mdx";
import ConfigureElementsSection from "@site/src/components/docs/_configure-elements-section.mdx";

# Collect Bank Accounts

> When building an e-commerce application, subscription service, payroll solution, or just need to enabling one-time purchases via direct debt (pay by bank) or fund transfers, one of the critical requirements is collecting and storing bank accounts information securely. However, bank routing numbers paired with account numbers, are considered sensitive data, and need to be securely collected in user forms and stored encrypted in special databases.

Reviewer: ```suggestion
When building an e-commerce application, subscription service, payroll solution, or just need to enable one-time purchases via direct debt (pay by bank) or fund transfers, one of the critical requirements is collecting and storing bank accounts information securely. However, bank routing numbers paired with account numbers, are considered sensitive data, and need to be securely collected in user forms and stored encrypted in special databases.
```

> When building an e-commerce application, subscription service, payroll solution, or just need to enabling one-time purchases via direct debt (pay by bank) or fund transfers, one of the critical requirements is collecting and storing bank accounts information securely. However, bank routing numbers paired with account numbers, are considered sensitive data, and need to be securely collected in user forms and stored encrypted in special databases.

Reviewer: what do you mean by `special` databases?

Author: Audited, secure, isolated. Maybe "special" was too generic. I will update the term.

> In this guide, we will set up Basis Theory SDKs to capture Bank Accounts in a frontend application, Web or Mobile, and securely store the data as tokens in Basis Theory Platform, [NACHA preferred partner for data tokenization and encryption](https://www.nacha.org/news/basis-theory-becomes-nacha-preferred-partner-data-tokenization-and-encryption). Given this guide is followed step by step, you can rest assured that you are complying to world's class security and privacy standards.

Reviewer: ```suggestion
In this guide, we will set up Basis Theory SDKs to capture Bank Accounts in a frontend, Web or Mobile application, and securely store the data as tokens in Basis Theory Platform, [NACHA's preferred partner for data tokenization and encryption](https://www.nacha.org/news/basis-theory-becomes-nacha-preferred-partner-data-tokenization-and-encryption). Given this guide is followed step by step, you can rest assured that you are complying to world's class security and privacy standards.
```

Author: In the article there is no apostrophe after "Nacha"

![Collect Bank Accounts Flowchart](/img/guides/banks/collect-banks-flowchart.png)

## Getting Started

<GettingStartedSection />

### Creating a Public Application

Next you will need a [Public Application](/docs/api/applications) using our NACHA-compliant template `Collect Bank Data`. [Click here to create one.](https://portal.basistheory.com/applications/create?application_template_id=d3ea55c1-4008-4703-9d52-db078ca5e245)

This will create an application with the following [Access Controls](/docs/concepts/access-controls):
* Permissions: `token:create`, `token:update`
* Containers: `/bank/`
* Transform: `mask`

<Alert>Save the API Key from the created Public Application as it will be used later in this guide.</Alert>

## Configuring Basis Theory Elements

<ConfigureElementsSection />

## Adding Text Elements

> Once installed and configured, add the Text Elements to your application. This will enable users to type in their bank account information in your form, while ensuring your systems never come in contact with the data.

Reviewer: ```suggestion
Once installed and configured, add the Text Elements to your application. This will enable users to type in their bank account information in your form, while ensuring your systems never touch the data.
```

Author: This would avoid repetition of the next occurrence. Thanks.

Notice how the bank `account_number`  has been masked before it is returned to your application. This default behavior prevents your application having to deal with the sensitive part of the bank details. [Click here](/docs/expressions/masks) to learn more about masking and see  how to customize this behavior.

## Customizing Tokens

The steps so far cover most cases when you need to collect bank accounts in your frontend and store them in a secure location. However, in some scenarios you may need to customize your bank tokens for specific business needs or technical requirements.

[Click here](/docs/concepts/what-are-tokens#what-are-the-capabilities-of-a-token) to learn more about Basis Theory Token capabilities.

### Deduplication

Companies often find it necessary to uniquely identify bank accounts flowing through their systems for various reasons, which may include: preventing fraudulent transactions, detecting abuse, building consumer profiles or streamlining payment processing for a better user experience.

> By leveraging [token fingerprinting](/docs/concepts/what-are-tokens#fingerprinting), developers can recognize the tokenized data in a customizable fashion, without having to touch with the plaintext data. For bank details, it is common to index in both Routing Number and Account Number, but in some cases additional consumer details may also be considered.

Reviewer: ```suggestion
By leveraging [token fingerprinting](/docs/concepts/what-are-tokens#fingerprinting), developers can recognize the tokenized data in a customizable fashion, without having to interact with the plaintext data. For bank details, it is common to index in both Routing Number and Account Number, but in some cases additional consumer details may also be considered.
```

    "restrictionPolicy": "mask"
  },
  "searchIndexes": [],
  "containers": [
    "/bank/high/"
  ]
}
```

If you want to prevent creation of a duplicate token based on the distinguishable fingerprint, add the flag below:

<Tabs className="bt-tabs" queryString="sdk">
  <TabItem value="javascript" label="JavaScript">

```javascript showLineNumbers title=index.js
async function submit () {
  try {
    const token = await bt.tokens.create({
      type: 'bank',
      data: {
        routing_number: routingNumberElement,
        account_number: accountNumberElement,
      },
      // highlight-start
      fingerprintExpression: '{{ data.account_number }}|{{ data.routing_number }}',
      deduplicateToken: true,
      // highlight-end
    });
  } catch (error) {
    console.error(error);
  }
}
```
  </TabItem>
  <TabItem value="react" label="React">

```jsx showLineNumbers title=App.jsx
export default function App() {
  const submit = async () => {
    try {
      const token = await bt?.tokens.create({
        type: 'bank',
        data: {
          routing_number: routingNumberElement,
          account_number: accountNumberElement,
        },
        // highlight-start

>         fingerprintExpression: '{{ data.account_number }}|{{ data.routing_number }}',

Reviewer: forgot flag

        // highlight-end
      });
    } catch (error) {
      console.error(error);
    }
  }

  return (...);
}
```
  </TabItem>
  <TabItem value="ios" label="iOS">

```swift showLineNumbers title=ViewController.swift
class ViewController: UIViewController {
    @IBAction func submit(_ sender: Any) {
        let body: [String: Any] = [
            "type": "bank",
            "data": [
                "routing_number": self.routingNumberTextField,
                "account_number": self.accountNumberTextField
            ],
            // highlight-start

>             "fingerprint_expression": "{{ data.account_number }}|{{ data.routing_number }}"

Reviewer: forgot flag

            // highlight-end
        ]

        BasisTheoryElements.tokenize(body: body, apiKey: "<API_KEY>") { token, error in
            guard error == nil else {
                print(error)
                return
            }
        }
    }
}
```

  </TabItem>
  <TabItem value="android" label="Android">

```kotlin showLineNumbers title=MainActivity.kt
class MainActivity : AppCompatActivity() {

    private fun submit() {
      val token = runBlocking {
        bt.tokens.create(object {
          val type = "bank"
          val data = object {
            val routing_number = routingNumberElement

>             val account_number = accountNumberElement

Reviewer: forgot flag

          }
          // highlight-start
          val fingerprint_expression = "{{ data.account_number }}|{{ data.routing_number }}"
          // highlight-end
        })
      }
    }
}
```
  </TabItem>
</Tabs>

By doing the above, you are instructing Basis Theory to return the existing token if it is found to have the same fingerprint. [Click here to learn more about token deduplication](/docs/concepts/what-are-tokens#deduplication).

## Conclusion

The best practices prescribed in this guide ensure that your user-facing applications can collect sensitive bank account information securely, while protected. The `token.id` obtained at the end of the [Storing Bank Accounts](#storing-bank-accounts) section is a synthetic replacement for the sensitive data and can be safely stored in your database, or transmitted through your systems, meeting security and privacy requirements while reducing the risk of exposure in case of data breaches.

> For next steps, take a look at the following guides to proceed taking the most value of your secured card tokens:

Reviewer: ```suggestion
For next steps, take a look at the following guides to proceed taking the most value of your secured bank tokens:
```
